### PR TITLE
format StmtAsyncWith

### PR DESCRIPTION
**Pull request by davidszotten**
refactor `FormatStmtWith` and share impl with `FormatStmtAsyncWith`

tests: snapshots

ref: #5368

### CODE REVIEW

File: crates/ruff_python_formatter/src/statement/stmt_async_with.rs [hunk 1]
```diff
@@ -1,12 +1,32 @@
...
+use ruff_python_ast::prelude::*;
 use rustpython_parser::ast::StmtAsyncWith;
 
+use super::stmt_with::FormatWithLike;
+
 #[derive(Default)]
 pub struct FormatStmtAsyncWith;
 
+impl<'ast> FormatWithLike<'ast> for StmtAsyncWith {
+    const ASYNC: bool = true;
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/src/statement/stmt_async_with.rs:12]: sorry, than continue doing that

**Review by MichaReiser — APPROVED** [review]

**Comment by davidszotten** [comment]
could you elaborate on what you mean by "to make ASYNC more explicit"?

(given that commit can do the same for try/trystar)

**Comment by MichaReiser** [comment]
Sure.

My main thinking is to use an `enum` if the implementations for the different types do not need to customize the formatting (significantly). Here, the only difference is the `async` and we can simply detect this by testing if the enum variant is `AsyncWith`. The main reason is that the `trait` doesn't integrate as nicely into the `Format` infrastructure because the trait cannot override `Format::fmt`, instead it is necessary to call the custom `fmt_<node>` method.

**Comment by davidszotten** [comment]
ok. implementing `Format` is nice (though this approach does add a fair amount of boilerplate, see e.g. https://github.com/astral-sh/ruff/pull/5418/files )

**Comment by MichaReiser** [comment]
> ok. implementing `Format` is nice (though this approach does add a fair amount of boilerplate, see e.g. [#5418 (files)](https://github.com/astral-sh/ruff/pull/5418/files) )

That's true!. Rome uses a `declare_node_union!` macro because declaring enums over a set of variants turned out to be an extremely valuable pattern. I think we could write something similar in the future to cut down on the boilerplate

https://github.com/rome/tools/blob/38104b339da8ff688f469799e1fc3f4a46f3d2ec/crates/rome_rowan/src/macros.rs#L47-L141

The rome version generates a bit more than what we need here but I found it super cool.

**Pull request merged**